src/udp.c

```diff
@@ -490,6 +490,98 @@ udp_close( udp_connection_t *uc )
   free(uc);
 }
 
+int
+udp_write_fill_source( udp_connection_t *uc, const void *buf, size_t len,
+                       struct sockaddr_storage *storage)
+{
+    int r;
+    struct sockaddr_in local_addr;
+    socklen_t local_addr_len = sizeof(local_addr);
+    struct sockaddr_in parent_addr;
+    socklen_t parent_addrlen = sizeof(parent_addr);
+    int reuse = 1;
+
+
+    if (storage == NULL)
+      storage = &uc->ip;
+
+    tvhdebug(uc->subsystem, "Got dst IP address: %s", inet_ntoa(((struct sockaddr_in*)storage)->sin_addr));
+
+    // Get the current socket configuration
+    if (getsockname(uc->fd, (struct sockaddr *)&parent_addr, &parent_addrlen) < 0) {
+        perror("getsockname");
+        return -1;
+    }
+    tvhdebug(uc->subsystem, "Got parent IP address: %s", inet_ntoa(parent_addr.sin_addr));
+
+    // Create a new socket
+    int cloned_sockfd = socket(AF_INET, SOCK_DGRAM, 0);
+    if (cloned_sockfd < 0) {
+        perror("socket");
+        return -1;
+    }
+
+
+    /* Mark reuse address */
+    if (setsockopt(cloned_sockfd, SOL_SOCKET, SO_REUSEADDR, &reuse, sizeof(reuse))) {
+      tvherror(uc->subsystem, "failed to reuse address for socket [%s]", strerror(errno));
+      close(cloned_sockfd);
+      return -1;
+    }
+
+    // Set the address to INADDR_ANY to let the routing table choose the source IP
+    parent_addr.sin_addr.s_addr = INADDR_ANY;
+
+    // Bind the new socket to the same port
+    if (bind(cloned_sockfd, (struct sockaddr *)&parent_addr, parent_addrlen) < 0) {
+        tvherror(uc->subsystem, "failed to bind address for socket [%s]", strerror(errno));
+        close(cloned_sockfd);
+        return -1;
+    }
+
+    // Connect the socket to the destination address
+    if (connect(cloned_sockfd, (struct sockaddr *)storage, sizeof(*storage)) < 0) {
+        tvherror(uc->subsystem, "connect() failed: %s", strerror(errno));
+        return -1;
+    }
+
+    // Get the local endpoint information (source IP address)
+    if (getsockname(cloned_sockfd, (struct sockaddr *)&local_addr, &local_addr_len) == -1) {
+        tvherror(uc->subsystem, "getsockname() failed: %s", strerror(errno));
+        return -1;
+    }
+
+    tvhdebug(uc->subsystem, "Got source IP address: %s", inet_ntoa(local_addr.sin_addr));
+
+    len += strlen(inet_ntoa(local_addr.sin_addr)) - 2;
+    char* data = malloc(len + 1);
+
+    len = snprintf(data, len + 1, buf, inet_ntoa(local_addr.sin_addr));
```

Comment:
Theres a bug here. 

snprintf returns "The number of characters that would have been written if n had been sufficiently large". This number can be larger than the size of 
`data` (and in my tests often is - I think because `buf` isnt a null terminated string). I think whats needed is to check that the return is >0 but ignore the specific value.

Alternatively, if `buf` were null-terminated, the return value should always be the same as `len`.

Reply:
This needs to be revisited anyways, as inet_ntoa will fail on IPv6, so this is not quite right yet and a few adjustments are still needed. I will keep this in mind, thanks!